lib/logstash/inputs/gelf.rb

```diff
@@ -267,11 +276,25 @@ def remap_gelf(event)
   end
 
   def strip_leading_underscore(event)
-     # Map all '_foo' fields to simply 'foo'
-     event.to_hash.keys.each do |key|
-       next unless key[0,1] == "_"
-       event.set(key[1..-1], event.get(key))
-       event.remove(key)
-     end
+    # Map all '_foo' fields to simply 'foo'
+    event.to_hash.keys.each do |key|
+      next unless key[0,1] == "_"
+      new_key = key[1..-1]
+      value = event.get(key)
+      if new_key == LogStash::Event::TIMESTAMP
+        event.set(new_key, coerce_timestamp_carefully(value))
+      else
+        event.set(new_key, value)
+      end
+      event.remove(key)
+    end
+  end
+
+  def coerce_timestamp_carefully(value)
+    LogStash::Timestamp.at(value)
+  rescue TypeError => e
```

Comment:
Can we possibly be less reactive, routing to BigDecimal if we have a reasonable expectation that doing so will succeed?

Reply:
So, if I understand correctly what you mean, instead of reaching a `TypeError` raised when `value` is a String and then fallback to `BigDecimal`, convert it directly to `BigDecimal` like:

```ruby
def coerce_timestamp_carefully(value)
  coerced_value = BigDecimal.new(value)
  LogStash::Timestamp.at(coerced_value.to_i, coerced_value.frac * 1000000)
end
```

Reply:
I updated with the suggestion, switched the regexp to accept also numbers in exponential notation like `0.123456e3`
adding the part:
```
0\.[0-9]+e[0-9]
```